contribute/design/1.x/cstor/20200511-cstor-backupandrestore.md

```diff
@@ -0,0 +1,259 @@
+- Solution to create an on-demand restore of backuped up cStor persistent volumes.
+- Solution to create a scheduled backup of cStor persistent volumes.
+- Solution to create an incremental/differential backup of cStor persistent volumes. With incremental/differential backup, the user will be able to save backup storage space and it 
+
+### Non-Goals
+
+- Supporting local backup and restore.
+
+## Proposal
+
+### User Stories
+
+#### Create a backup for CStorVolumes
+As an OpenEBS user, I should be able create a backup for CStorVolumes.
+
+#### Create a restore for backuped CStorVolumes
+As an OpenEBS user, I should be restore a backuped CStorVolumes.
+
+#### Scheduled backup of CStorVolumes
+As an OpenEBS user, I should be able to create a scheduled backups for CStorVolumes.
+
+### Proposed Implementation
+
+#### OpenEBS Velero Plugin
+
+TODO: @vishnuitta to fill plugin side approach.
+
+#### CVC-Operator REST Interface
+
+The volume management api like create, delete, snapshot, clone etc have moved from m-apiserver to CVC-Operator as part of supporting the CSI Driver. The backup/restore of volume API also should be moved to the CVC Operator as the backup/restore api in turn depends on the volume snapshot, clone implementations. Currently CVC Operator only supports declarative API via CRs, but plugin requires imperative API(REST API). The proposal is to implement a REST server within CVC-Operator to perform backup/restore operations. The CStor backup/restore module should identify the type of the volume and forward the REST API requests to the CVC Operator.
+
+As mentioned [above](#openebs-veleroplugin) velero-plugin triggers different REST API calls based on the type of request from velero-server.
+
+### Steps to perform user stories
+
+- User can create backup using velero CLI
+  Ex: velero backup create <BACKUP_NAME> --include-namespaces=<NAME_SPACE> --snapshot-volumes –volume-snapshot-locations=<SNAPSHOT_LOCATION>
+- User can resotre backuped using velero CLI
+  Ex: velero restore create --from-backup <BACKUP_NAME> --restore-volumes=true
+
+### Low Level Design
+
+#### Work Flow
+
+##### Velero Server sends CreateSnapshot Request
+
+Velero server sends CreateSnapshot API with volumeID to create the snapshot. Velero Interface will `CreateSnapshot` API of the cStor backup/restore controller. cStor backup/restore controller is responsible for executing below steps:
+1. BackUp the relevant PVC object to the cloud provider.
+2. Execute the REST API(POST `/latest/backups`) of CVC-Operator (This object will include the IP Address of snapshot receiver/sender module) to create backup.
+    2.1 Create a snapshot using volume name and snapshot name(which will get during the CreateSnapshot request).
+    2.2 Find the Healthy cStorVolumeReplica if it doesn't exist return error.
+    2.3 Create CStorCompleteBackUp resources if it doesn’t exist(Intention of creating this resource is used for incremental backup purpose).
```

Comment:
```suggestion
    2.3 Create CStorCompletedBackUp resources if it doesn’t exist(Intention of creating this resource is used for incremental backup purpose).
```

Reply:
same for other places also wherever it is used

Reply:
Updated in all other places also